### DISCUSSION TITLE
Suggestions for the MerkleStreamer API

### DISCUSSION THREAD

**Discussion by razgraf** [sablier-labs/v2-periphery#189] [Ideas]
### Parameters for the factory `create` method

https://github.com/sablier-labs/v2-periphery/blob/89aa8bd85729d169fbf43b53f38514f82f42e53f/src/SablierV2MerkleStreamerFactory.sol#L50-L51

**Question:** Why is `campaignTotalAmount` prefixed with `campaign` while `recipientsCount` isn't?
**Suggestion:** Could we somehow bundle these parameters, either by prefixing them similarly or (better) by grouping them under an attributes struct? I'm mostly thinking of those that are meant to be emitted for indexing, so: `[CID, amount, recipients]`. Alternatively, simply removing the campaign prefix would also be much less verbose.

### Consistency in parameter order

https://github.com/sablier-labs/v2-periphery/blob/89aa8bd85729d169fbf43b53f38514f82f42e53f/src/SablierV2MerkleStreamerFactory.sol#L57-L62

I don't think it influences the salt in a bad way so could we maybe swap the order of `lockupLinear` and `initialAdmin` to mirror the one we use below for the new LL constructor?

**Comment by andreivladbrg** [comment]
> Why is campaignTotalAmount prefixed with campaign while recipientsCount isn't?

It was done to avoid any potential confusion with the `totalAmount` parameter from the create function:

https://github.com/sablier-labs/v2-periphery/blob/89aa8bd85729d169fbf43b53f38514f82f42e53f/src/SablierV2MerkleStreamerLL.sol#L92

We could add "campaign" prefix on `recipientsCount` as well, or just remove it. Both work for me. @PaulRBerg  you decide 😅

> [CID, amount, recipients] 

So, are you suggesting something like this?
```solidity
    struct EmitParameters {
        string ipfsCID;
        uint256 campaignTotalAmount,
        uint256 campaignRecipientsCount
    }

    function createMerkleStreamerLL(
        address initialAdmin,
        ISablierV2LockupLinear lockupLinear,
        IERC20 asset,
        bytes32 merkleRoot,
        uint40 expiration,
        LockupLinear.Durations memory streamDurations,
        bool cancelable,
        EmitParameters memory emitParameters
    )  external
       returns (ISablierV2MerkleStreamerLL merkleStreamerLL);
```

>I don't think it influences the salt in a bad way so could we maybe swap the order of lockupLinear and initialAdmin to mirror the one we use below for the new LL constructor?

Good idea, we could do that.

**Comment by PaulRBerg** [comment]
> Why is `campaignTotalAmount` prefixed with campaign while `recipientsCount` isn't?

@andreivladbrg answered this. Each stream's deposit amount is called `totalAmount` in V2 Core, so we added the "campaign" prefix as a qualifier to differentiate it from the stream amounts.

Rather than adding a superfluous prefix to `recipientsCount` (which doesn't need any "campaign" qualifier), what if we renamed `campaignTotalAmount` to `aggregateAmount`?

> grouping them under an attributes struct?

Structs have bad UX for integrations (see: Tally integration on Telegram).

I suggest avoiding structs unless they are necessary (which isn't the case here because there are no StackTooDeep issues).

> swap the order of `lockupLinear` and `initialAdmin`

sounds good - @andreivladbrg can you implement this change, please
